hfs_fit.py

```diff
@@ -0,0 +1,717 @@
+import numpy as np
```

Comment:
first up, a 700 line file with this level of cross dependencies is reeeeally tough to read, could we consider splitting it out somehow if we can